File: 0764. N-ary Tree Level Order Traversal/Solution04.py
/*
Submission Detail:{
    Difficulty : Medium
    Acceptance Rate : 70.51 %
    Runtime : 71 ms
    Memory Usage : 16.2 MB
    Testcase : 38 / 38 passed
    Ranking : 
        Your runtime beats 49.10 % of python submissions.
        Your memory usage beats 98.52 % of python submissions.
}
*/

"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution(object):
    def levelOrder(self, root):
        listtoreturn=[]
        q= collections.deque()
        q.append(root)
        while q:
            Qlen = len(q)
            level = []
            for i in range(Qlen):
                node=q.popleft()
                if node:
                    level.append(node.val)
                    for c in node.children:
                        q.append(c)
            if level:
                listtoreturn.append(level)
        return listtoreturn

File: 0662. Maximum Width of Binary Tree/Solution04.java
/*
Submission Detail:{
    Difficulty : Medium
    Acceptance Rate : 40.49 %
    Runtime : 2 ms
    Memory Usage : 42.9 MB
    Testcase : 114 / 114 passed
    Ranking : 
        Your runtime beats 89.55 % of java submissions.
        Your memory usage beats 84.90 % of java submissions.
}
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int widthOfBinaryTree(TreeNode root) {
        List<Integer> lefts = new ArrayList<Integer>(); // left most nodes at each level;
        int[] res = new int[1]; // max width
        dfs(root, 1, 0, lefts, res);
        return res[0];
    }
    private void dfs(TreeNode node, int id, int depth, List<Integer> lefts, int[] res) {
        if (node == null) return;
        if (depth >= lefts.size()) lefts.add(id);   // add left most node
        res[0] = Integer.max(res[0], id + 1 - lefts.get(depth));
        dfs(node.left,  id * 2,     depth + 1, lefts, res);
        dfs(node.right, id * 2 + 1, depth + 1, lefts, res);
    }
}